File: payment-service/src/main/java/com/hedvig/paymentservice/services/segmentPublisher/EventListener.java
package com.hedvig.paymentservice.services.segmentPublisher;

import com.google.common.collect.ImmutableMap;
import com.hedvig.paymentservice.domain.payments.events.TrustlyAccountCreatedEvent;
import com.segment.analytics.Analytics;
import com.segment.analytics.messages.IdentifyMessage;
import java.util.Map;
import lombok.val;
import org.axonframework.eventhandling.EventHandler;
import org.springframework.stereotype.Component;

@Component
public class EventListener {

    private final Analytics segmentAnalytics;

    public EventListener(Analytics analytics) {
        this.segmentAnalytics = analytics;
    }

    @EventHandler
    public void on(TrustlyAccountCreatedEvent evt) {


        ImmutableMap<String, Object> traits = ImmutableMap.of("is_direct_debit_activated", evt.isDirectDebitMandateActivated());

        segmentIdentify(traits, evt.getMemberId());

    }

    private void segmentIdentify(Map<String, Object> traitsMap, String memberId) {
        segmentAnalytics.enqueue(
            IdentifyMessage.builder()
                .userId(memberId)
                .enableIntegration("All", false)
                .enableIntegration("Customer.io", true)
                .traits(traitsMap));
    }

}

File: payment-service/src/main/java/com/hedvig/paymentservice/services/segmentPublisher/EventListener.kt
package com.hedvig.paymentservice.services.segmentPublisher

import com.hedvig.paymentservice.domain.payments.enums.AdyenAccountStatus
import com.hedvig.paymentservice.domain.payments.events.AdyenAccountCreatedEvent
import com.hedvig.paymentservice.domain.payments.events.AdyenAccountUpdatedEvent
import com.hedvig.paymentservice.domain.payments.events.DirectDebitConnectedEvent
import com.hedvig.paymentservice.domain.payments.events.DirectDebitDisconnectedEvent
import com.hedvig.paymentservice.graphQl.types.PayinMethodStatus
import com.hedvig.paymentservice.serviceIntergration.notificationService.NotificationService
import org.axonframework.config.ProcessingGroup
import org.axonframework.eventhandling.EventHandler
import org.springframework.context.annotation.Profile
import org.springframework.stereotype.Component

@Component
@Profile("customer.io")
@ProcessingGroup("SegmentProcessorGroup")
class EventListener(
    private val notificationService: NotificationService
) {
    @EventHandler
    fun on(evt: DirectDebitConnectedEvent) {
        val traits = mapOf("is_direct_debit_activated" to true)

        notificationService.updateCustomer(evt.memberId, traits)
    }

    @EventHandler
    fun on(evt: DirectDebitDisconnectedEvent) {
        val traits = mapOf("is_direct_debit_activated" to false)

        notificationService.updateCustomer(evt.memberId, traits)
    }

    @EventHandler
    fun on(e: AdyenAccountCreatedEvent) {
        updateTraitsBasedOnAdyenAccountStatus(e.memberId, e.accountStatus)
    }

    @EventHandler
    fun on(e: AdyenAccountUpdatedEvent) {
        updateTraitsBasedOnAdyenAccountStatus(e.memberId, e.accountStatus)
    }

    private fun updateTraitsBasedOnAdyenAccountStatus(memberId: String, adyenAccountStatus: AdyenAccountStatus) {
        val traits = when (PayinMethodStatus.fromAdyenAccountStatus(adyenAccountStatus)) {
            PayinMethodStatus.ACTIVE -> mapOf(IS_CARD_CONNECTED to true)
            PayinMethodStatus.PENDING, PayinMethodStatus.NEEDS_SETUP -> mapOf(IS_CARD_CONNECTED to false)
        }
        notificationService.updateCustomer(memberId, traits)
    }

    companion object {
        const val IS_CARD_CONNECTED: String = "is_card_connected"
    }
}
